File: meta_ads/campaigns.py
"""Kampanya / ad set / reklam olusturma ve mevcut kampanya yonetimi.

Guvenlik kurali: `create_paused_campaign` her zaman PAUSED durumda kampanya/
adset/reklam olusturur. ENABLED'a gecis sadece `resume()` ile, ayri ve acik
bir adimda yapilir - burada asla otomatik tetiklenmez.
"""
from __future__ import annotations

from typing import Any, Optional

from facebook_business.adobjects.ad import Ad
from facebook_business.adobjects.adset import AdSet
from facebook_business.adobjects.campaign import Campaign

from . import creatives
from .audiences import ensure_ready
from .client import call, ensure_api_initialized, get_ad_account

# NOT: Meta bu enum degerlerini zaman zaman gunceller/yeniden adlandirir.
# Ilk gercek kampanyadan once Meta Ads Manager > Kampanya olustur akisinda
# ayni objective/optimization_goal degerlerinin hala gecerli oldugunu teyit et.
DEFAULT_OBJECTIVE = "OUTCOME_APP_PROMOTION"
DEFAULT_OPTIMIZATION_GOAL = "APP_INSTALLS"
DEFAULT_BILLING_EVENT = "IMPRESSIONS"
DEFAULT_COUNTRIES = ["TR"]


def _try_to_cents(try_amount: float) -> int:
    """Turk Lirasi'ni Meta'nin bekledigi en kucuk birime (kurus) cevirir."""
    return int(round(try_amount * 100))


def build_campaign_payload(*, name: str, objective: str = DEFAULT_OBJECTIVE) -> dict[str, Any]:
    return {
        Campaign.Field.name: name,
        Campaign.Field.objective: objective,
        Campaign.Field.status: Campaign.Status.paused,
        "special_ad_categories": [],
        # Butcemiz her zaman ad-set seviyesinde (CBO kapali) - bu araç
        # kampanya seviyesinde butce kullanmiyor. Meta bu alani artik
        # zorunlu kiliyor (asagidaki False secimi olmadan create_campaign
        # "is_adset_budget_sharing_enabled alaninda True veya False
        # belirtilmelidir" hatasi veriyor).
        "is_adset_budget_sharing_enabled": False,
    }


def build_adset_payload(
    *,
    name: str,
    campaign_id: str,
    daily_budget_try: float,
    audience_id: Optional[str] = None,
    app_id: str,
    app_store_url: str,
    optimization_goal: str = DEFAULT_OPTIMIZATION_GOAL,
    billing_event: str = DEFAULT_BILLING_EVENT,
    countries: Optional[list[str]] = None,
    custom_event_type: Optional[str] = None,
) -> dict[str, Any]:
    promoted_object: dict[str, Any] = {
        "application_id": app_id,
        "object_store_url": app_store_url,
    }
    # custom_event_type verilirse (ornegin "SUBSCRIBE"), optimizasyon App
    # Installs yerine belirli bir uygulama-ici event'i hedefler. Kitle zaten
    # uygulamayi yuklemis kisilerden olusuyorsa (ornegin "acik ama abone
    # degil") bu sart - App Installs optimizasyonu bu kisiler icin anlamsiz.
    if custom_event_type:
        promoted_object["custom_event_type"] = custom_event_type

    targeting: dict[str, Any] = {
        "geo_locations": {"countries": countries or DEFAULT_COUNTRIES},
        # object_store_url her zaman Apple App Store'a isaret ediyor -
        # targeting bunu iOS'a kisitlamazsa Meta "Mobile Targeting
        # Mismatch" hatasi veriyor (uygulama tek platform, hedefleme
        # coklu platform varsayiyordu).
        "user_os": ["iOS"],
    }
    # audience_id verilmezse Meta'nin genis/Advantage+ hedeflemesine
    # birakilir (orn. mevcut "claude TOF" kampanyasi da custom_audiences
    # kullanmiyor).
    if audience_id:
        targeting["custom_audiences"] = [{"id": audience_id}]

    return {
        AdSet.Field.name: name,
        AdSet.Field.campaign_id: campaign_id,
        AdSet.Field.daily_budget: _try_to_cents(daily_budget_try),
        AdSet.Field.billing_event: billing_event,
        AdSet.Field.optimization_goal: optimization_goal,
        # Meta artik teklif stratejisinin acikca belirtilmesini istiyor.
        # LOWEST_COST_WITHOUT_CAP = teklif tutari/sinir gerektirmeyen, en
        # dusuk maliyeti hedefleyen varsayilan strateji.
        AdSet.Field.bid_strategy: AdSet.BidStrategy.lowest_cost_without_cap,
        AdSet.Field.status: AdSet.Status.paused,
        AdSet.Field.promoted_object: promoted_object,
        AdSet.Field.targeting: targeting,
    }


def build_ad_payload(*, name: str, adset_id: str, creative_id: str) -> dict[str, Any]:
    return {
        Ad.Field.name: name,
        Ad.Field.adset_id: adset_id,
        Ad.Field.creative: {"creative_id": creative_id},
        Ad.Field.status: Ad.Status.paused,
    }


def create_paused_campaign(
    *,
    name: str,
    daily_budget_try: float,
    app_id: str,
    app_store_url: str,
    page_id: str,
    link: str,
    creative_type: str,  # "single" | "carousel"
    images: list[str],
    headlines: list[str],
    descriptions: Optional[list[str]] = None,
    message: str = "",
    objective: str = DEFAULT_OBJECTIVE,
    optimization_goal: str = DEFAULT_OPTIMIZATION_GOAL,
    billing_event: str = DEFAULT_BILLING_EVENT,
    countries: Optional[list[str]] = None,
    custom_event_type: Optional[str] = None,
    audience_id: Optional[str] = None,
    dry_run: bool = False,
) -> dict[str, Any]:
    if creative_type not in ("single", "carousel"):
        raise ValueError("creative_type 'single' ya da 'carousel' olmali.")
    if creative_type == "carousel" and len(headlines) != len(images):
        raise ValueError("Carousel'de her gorsel icin bir headline gerekli (sayilar esit degil).")
    if creative_type == "single" and len(images) != 1:
        raise ValueError("Tekil (single) kreatif icin tam olarak 1 gorsel gerekli.")

    # Kitle hazir degilse burada durur - hicbir API cagrisi yapilmadan once.
    # audience_id verilmezse bu kontrol atlanir (genis/Advantage+ hedefleme).
    if audience_id:
        ensure_ready(audience_id)

    campaign_payload = build_campaign_payload(name=name, objective=objective)
    adset_payload_preview = build_adset_payload(
        name=f"{name} - adset",
        campaign_id="<olusturulacak>",
        daily_budget_try=daily_budget_try,
        audience_id=audience_id,
        app_id=app_id,
        app_store_url=app_store_url,
        optimization_goal=optimization_goal,
        billing_event=billing_event,
        countries=countries,
        custom_event_type=custom_event_type,
    )

    plan = {
        "campaign": campaign_payload,
        "adset": adset_payload_preview,
        "creative_type": creative_type,
        "images": images,
        "headlines": headlines,
        "descriptions": descriptions,
        "link": link,
        "page_id": page_id,
    }

    if dry_run:
        return {"dry_run": True, "plan": plan}

    account = get_ad_account()

    campaign = call(account.create_campaign, params=campaign_payload)
    campaign_id = campaign[Campaign.Field.id]

    adset_payload = build_adset_payload(
        name=f"{name} - adset",
        campaign_id=campaign_id,
        daily_budget_try=daily_budget_try,
        audience_id=audience_id,
        app_id=app_id,
        app_store_url=app_store_url,
        optimization_goal=optimization_goal,
        billing_event=billing_event,
        countries=countries,
        custom_event_type=custom_event_type,
    )
    adset = call(account.create_ad_set, params=adset_payload)
    adset_id = adset[AdSet.Field.id]

    image_hashes = [creatives.upload_image(path) for path in images]

    if creative_type == "single":
        creative_payload = creatives.build_single_image_creative(
            page_id=page_id,
            link=link,
            image_hash=image_hashes[0],
            message=message,
            headline=headlines[0],
        )
    else:
        descs = descriptions or [""] * len(images)
        cards = [
            {"image_hash": h, "name": headline, "description": desc}
            for h, headline, desc in zip(image_hashes, headlines, descs)
        ]
        creative_payload = creatives.build_carousel_creative(
            page_id=page_id, link=link, message=message, cards=cards
        )

    creative = call(account.create_ad_creative, params=creative_payload)
    creative_id = creative["id"]

    ad_payload = build_ad_payload(name=f"{name} - ad", adset_id=adset_id, creative_id=creative_id)
    ad = call(account.create_ad, params=ad_payload)
    ad_id = ad[Ad.Field.id]

    return {
        "dry_run": False,
        "campaign_id": campaign_id,
        "adset_id": adset_id,
        "creative_id": creative_id,
        "ad_id": ad_id,
        "status": "PAUSED",
    }


def _set_campaign_status(campaign_id: str, status: str) -> None:
    ensure_api_initialized()
    campaign = Campaign(campaign_id)
    call(campaign.api_update, params={Campaign.Field.status: status})


def pause(campaign_id: str) -> None:
    _set_campaign_status(campaign_id, Campaign.Status.paused)


def resume(campaign_id: str) -> None:
    """DIKKAT: Bu kampanyayi ACTIVE yapar, harcama baslar. cli.py onay ister."""
    _set_campaign_status(campaign_id, Campaign.Status.active)


def set_budget(campaign_id: str, daily_budget_try: float) -> None:
    """Kampanya seviyesinde (CBO) gunluk butceyi gunceller.

    NOT: Eger kampanya ad-set seviyesinde butce kullaniyorsa (CBO kapali),
    bu cagri etkisiz kalir - o durumda AdSet.daily_budget guncellenmeli.
    """
    ensure_api_initialized()
    campaign = Campaign(campaign_id)
    call(
        campaign.api_update,
        params={Campaign.Field.daily_budget: _try_to_cents(daily_budget_try)},
    )


def get_campaign_status(campaign_id: str) -> dict[str, Any]:
    ensure_api_initialized()
    campaign = Campaign(campaign_id)
    fields = [Campaign.Field.name, Campaign.Field.status, Campaign.Field.daily_budget, Campaign.Field.objective]
    data = call(campaign.api_get, fields=fields)
    return dict(data)

File: worker/src/meta/campaigns.ts
/**
 * Kampanya / ad set / reklam olusturma ve mevcut kampanya yonetimi.
 * meta_ads/campaigns.py'nin portu.
 *
 * GUVENLIK KURALI: `createPausedCampaign` her zaman PAUSED durumda kampanya/
 * adset/reklam olusturur. ACTIVE'e gecis bu dosyada YOK - MCP tool katmani
 * (src/mcp-agent.ts) resume icin ayri bir preview+confirm_token akisi
 * kullaniyor, tek cagriyla asla harcama baslamaz.
 *
 * Asagidaki sabitler bu oturumda (2026-09-01/02) gercek Meta API'ye karsi
 * canli test edilerek bulundu - Meta bunlari artik zorunlu kiliyor:
 *   - is_adset_budget_sharing_enabled: false  (CBO kullanilmiyorsa sart)
 *   - bid_strategy: LOWEST_COST_WITHOUT_CAP    (teklif stratejisi artik acik olmali)
 *   - targeting.user_os: ["iOS"]               (object_store_url iOS'a isaret ediyorsa sart)
 * Ayrica: minimum ad-set gunluk butcesi hesaba gore degisir (bu oturumda
 * act_244832992826003 icin ~48.23 TRY idi) - dusuk deger acik bir hata doner.
 */
import { adAccountPath, graphRequest, graphRequestPaged, type MetaEnv } from "./client";
import { ensureAudienceReady } from "./audiences";
import { buildCreativeForType, type ImageInput } from "./creatives";

export const DEFAULT_OBJECTIVE = "OUTCOME_APP_PROMOTION";
export const DEFAULT_OPTIMIZATION_GOAL = "APP_INSTALLS";
export const DEFAULT_BILLING_EVENT = "IMPRESSIONS";
export const DEFAULT_COUNTRIES = ["TR"];

function tryToCents(tryAmount: number): number {
  return Math.round(tryAmount * 100);
}

export function buildCampaignPayload(opts: { name: string; objective?: string }) {
  return {
    name: opts.name,
    objective: opts.objective ?? DEFAULT_OBJECTIVE,
    status: "PAUSED",
    special_ad_categories: [] as string[],
    is_adset_budget_sharing_enabled: false,
  };
}

export function buildAdSetPayload(opts: {
  name: string;
  campaignId: string;
  dailyBudgetTry: number;
  audienceId?: string;
  appId: string;
  appStoreUrl: string;
  optimizationGoal?: string;
  billingEvent?: string;
  countries?: string[];
  customEventType?: string;
}) {
  const promotedObject: Record<string, unknown> = {
    application_id: opts.appId,
    object_store_url: opts.appStoreUrl,
  };
  if (opts.customEventType) promotedObject.custom_event_type = opts.customEventType;

  const targeting: Record<string, unknown> = {
    geo_locations: { countries: opts.countries ?? DEFAULT_COUNTRIES },
    user_os: ["iOS"],
  };
  // audience_id verilmezse Meta'nin genis/Advantage+ hedeflemesine birakilir
  // (orn. mevcut "claude TOF" kampanyasi da custom_audiences kullanmiyor).
  if (opts.audienceId) {
    targeting.custom_audiences = [{ id: opts.audienceId }];
  }

  return {
    name: opts.name,
    campaign_id: opts.campaignId,
    daily_budget: tryToCents(opts.dailyBudgetTry),
    billing_event: opts.billingEvent ?? DEFAULT_BILLING_EVENT,
    optimization_goal: opts.optimizationGoal ?? DEFAULT_OPTIMIZATION_GOAL,
    bid_strategy: "LOWEST_COST_WITHOUT_CAP",
    status: "PAUSED",
    promoted_object: promotedObject,
    targeting,
  };
}

export function buildAdPayload(opts: { name: string; adsetId: string; creativeId: string }) {
  return {
    name: opts.name,
    adset_id: opts.adsetId,
    creative: { creative_id: opts.creativeId },
    status: "PAUSED",
  };
}

export interface CreateCampaignInput {
  name: string;
  dailyBudgetTry: number;
  /** Verilmezse Meta'nin genis/Advantage+ hedeflemesine birakilir (custom audience kullanilmaz). */
  audienceId?: string;
  appId: string;
  appStoreUrl: string;
  pageId: string;
  link: string;
  creativeType: "single" | "carousel" | "video";
  images?: ImageInput[];
  video?: ImageInput;
  thumbnail?: ImageInput;
  headlines: string[];
  descriptions?: string[];
  message?: string;
  objective?: string;
  optimizationGoal?: string;
  billingEvent?: string;
  countries?: string[];
  customEventType?: string;
  dryRun?: boolean;
}

export async function createPausedCampaign(
  env: MetaEnv & { CREATIVES?: R2Bucket },
  input: CreateCampaignInput,
) {
  if (!["single", "carousel", "video"].includes(input.creativeType)) {
    throw new Error("creative_type 'single', 'carousel' ya da 'video' olmali.");
  }

  // Kitle hazir degilse burada durur - hicbir yazma cagrisi yapilmadan once.
  // audience_id verilmezse bu kontrol atlanir (genis/Advantage+ hedefleme).
  if (input.audienceId) {
    await ensureAudienceReady(env, input.audienceId);
  }

  const campaignPayload = buildCampaignPayload({ name: input.name, objective: input.objective });
  const adsetPreview = buildAdSetPayload({
    name: `${input.name} - adset`,
    campaignId: "<olusturulacak>",
    dailyBudgetTry: input.dailyBudgetTry,
    audienceId: input.audienceId,
    appId: input.appId,
    appStoreUrl: input.appStoreUrl,
    optimizationGoal: input.optimizationGoal,
    billingEvent: input.billingEvent,
    countries: input.countries,
    customEventType: input.customEventType,
  });

  const plan = {
    campaign: campaignPayload,
    adset: adsetPreview,
    creative_type: input.creativeType,
    images: input.images,
    headlines: input.headlines,
    descriptions: input.descriptions ?? null,
    link: input.link,
    page_id: input.pageId,
  };

  if (input.dryRun) {
    return { dry_run: true, plan };
  }

  const campaign = await graphRequest(env, adAccountPath(env, "campaigns"), {
    method: "POST",
    params: campaignPayload,
  });
  const campaignId = campaign.id as string;

  const adsetPayload = buildAdSetPayload({
    name: `${input.name} - adset`,
    campaignId,
    dailyBudgetTry: input.dailyBudgetTry,
    audienceId: input.audienceId,
    appId: input.appId,
    appStoreUrl: input.appStoreUrl,
    optimizationGoal: input.optimizationGoal,
    billingEvent: input.billingEvent,
    countries: input.countries,
    customEventType: input.customEventType,
  });
  const adset = await graphRequest(env, adAccountPath(env, "adsets"), {
    method: "POST",
    params: adsetPayload,
  });
  const adsetId = adset.id as string;

  const creativePayload = await buildCreativeForType(env, {
    creativeType: input.creativeType,
    images: input.images,
    video: input.video,
    thumbnail: input.thumbnail,
    headlines: input.headlines,
    descriptions: input.descriptions,
    message: input.message,
    link: input.link,
    pageId: input.pageId,
  });

  const creative = await graphRequest(env, adAccountPath(env, "adcreatives"), {
    method: "POST",
    params: creativePayload,
  });
  const creativeId = creative.id as string;

  const adPayload = buildAdPayload({ name: `${input.name} - ad`, adsetId, creativeId });
  const ad = await graphRequest(env, adAccountPath(env, "ads"), {
    method: "POST",
    params: adPayload,
  });

  return {
    dry_run: false,
    campaign_id: campaignId,
    adset_id: adsetId,
    creative_id: creativeId,
    ad_id: ad.id as string,
    status: "PAUSED" as const,
  };
}

export async function listCampaigns(
  env: MetaEnv,
  opts: { status?: string[]; limit?: number } = {},
): Promise<Record<string, unknown>[]> {
  const fields = ["id", "name", "status", "objective", "daily_budget"].join(",");
  const params: Record<string, unknown> = { fields, limit: opts.limit ?? 50 };
  if (opts.status?.length) params.effective_status = opts.status;
  return await graphRequestPaged(env, adAccountPath(env, "campaigns"), params);
}

export async function pauseCampaign(env: MetaEnv, campaignId: string): Promise<void> {
  await graphRequest(env, `/${campaignId}`, { method: "POST", params: { status: "PAUSED" } });
}

/** DIKKAT: Bu kampanyayi ACTIVE yapar, harcama baslar. Tool katmani onay ister. */
export async function resumeCampaign(env: MetaEnv, campaignId: string): Promise<void> {
  await graphRequest(env, `/${campaignId}`, { method: "POST", params: { status: "ACTIVE" } });
}

/**
 * Kampanya (CBO) gunluk butcesini gunceller. NOT: bu proje olusturdugu her
 * kampanyada butceyi ad-set seviyesinde ayarliyor (CBO kapali) - bu fonksiyon
 * o kampanyalarda etkisiz kalir. Gercek degisiklik icin meta/adsets.ts'deki
 * setAdSetBudget kullanilmali. Bu fonksiyon sadece CBO acik, disaridan/elle
 * olusturulmus kampanyalar icin anlamli.
 */
export async function setCampaignBudget(
  env: MetaEnv,
  campaignId: string,
  dailyBudgetTry: number,
): Promise<void> {
  await graphRequest(env, `/${campaignId}`, {
    method: "POST",
    params: { daily_budget: tryToCents(dailyBudgetTry) },
  });
}

export async function getCampaignStatus(env: MetaEnv, campaignId: string): Promise<Record<string, unknown>> {
  const fields = ["name", "status", "daily_budget", "objective"].join(",");
  return await graphRequest(env, `/${campaignId}`, { params: { fields } });
}

/** DIKKAT: Kalicidir, geri alinamaz. Tool katmani sadece PAUSED kampanyalar icin ve onayla izin verir. */
export async function deleteCampaign(env: MetaEnv, campaignId: string): Promise<void> {
  await graphRequest(env, `/${campaignId}`, { method: "DELETE" });
}
